pysd/py_backend/output.py

```diff
@@ -0,0 +1,513 @@
+
+        Parameters
+        ----------
+        model: pysd.Model
+            PySD Model object
+        capture_elements: set
+            Which model elements to capture - uses pysafe names.
+
+        Returns
+        -------
+        None
+
+        """
+        self.ds = nc.Dataset(self.out_file, "w")
+
+        # defining global attributes
+        self.ds.description = "Results for simulation run on " \
+            f"{t.ctime(t.time())} using PySD version {__version__}"
+        self.ds.model_file = model.py_model_file or model.mdl_file
+        self.ds.timestep = f"{model.time.time_step()}" if model.cache_type[
+            "time_step"] == "run" else "Variable"
+        self.ds.initial_time = f"{model.time.initial_time()}"
+        self.ds.final_time = f"{model.time.final_time()}" if model.cache_type[
+            "final_time"] == "run" else "Variable"
+
+        # creating variables for all model dimensions
+        for dim_name, coords in model.subscripts.items():
+            coords = np.array(coords)
+            # create dimension
+            self.ds.createDimension(dim_name, len(coords))
+
+            # length of the longest string in the coords
+            max_str_len = len(max(coords, key=len))
+
+            # create variable for the dimension
+            var = self.ds.createVariable(
+                dim_name, f"S{max_str_len}", (dim_name,))
+
+            # assigning coords to dimension
+            var[:] = coords
+
+        # creating the time dimension as unlimited
+        self.ds.createDimension("time", None)
+
+        # creating variables
+        self.__create_ds_vars(model, capture_elements)
+
+    def update(self, model, capture_elements):
+        """
+        Writes values of variables in capture_elements in the netCDF4 Dataset.
+
+        Parameters
+        ----------
+        model: pysd.Model
+            PySD Model object
+        capture_elements: set
+            Which model elements to capture - uses pysafe names.
+
+        Returns
+        -------
+        None
+        """
+        for key in capture_elements:
+
+            comp = model[key]
+
+            if "time" in self.ds[key].dimensions:
+                if isinstance(comp, xr.DataArray):
+                    self.ds[key][self.step, :] = comp.values
+                else:
+                    self.ds[key][self.step] = comp
+            else:
+                try:  # this issue can arise with external objects
+                    if isinstance(comp, xr.DataArray):
+                        self.ds[key][:] = comp.values
+                    else:
+                        self.ds[key][:] = comp
+                except ValueError:
+                    warnings.warn(
+                        f"The dimensions of {key} in the results "
+                        "do not match the declared dimensions for this "
+                        "variable. The resulting values will not be "
+                        "included in the results file.")
```

Comment:
Could you split the update method in two?
For me it doesn't make sense to execute always the if/else as the constant values and time-dependent values are included in different parts of the workflow

The ValueError is not tested, however it should never be raised but because of the bug of the external data. Please, open an issue about that bug. I would remove this error then.

Reply:
Functions have been split.
Issue #363 has been created and the try/except block removed from the code.